src/visualizers/VisualizerFactorHistogram.ts

```diff
@@ -0,0 +1,200 @@
+                        counter += Number(power)
+                    } else if (base === 0n) {
+                        counter = 0
+                    }
+                }
+            }
+
+            factorArray[i] = counter
+        }
+        return factorArray
+    }
+
+    // Create an array with the frequency of each number
+    // of factors in the corresponding bins
+    binFactorArray(): number[] {
+        const binFactorArray = []
+        const factorArray = this.factorArray()
+        for (
+            let i = 0;
+            i < natlog(this.largestValue()) / natlog(2) + 1;
+            i++
+        ) {
+            binFactorArray.push(0)
+        }
+
+        for (let i = 0; i < factorArray.length; i++) {
+            binFactorArray[Math.trunc(factorArray[i] / this.binSize)]++
+        }
+
+        return binFactorArray
+    }
+
+    // Create a number that represents how
+    // many pixels wide each bin should be
+    binWidth(): number {
+        const binFactorArray = this.binFactorArray()
+        let array_length = 0
+        for (let i = 0; i < binFactorArray.length; i++) {
+            if (binFactorArray[i] != 0) {
+                array_length = i
+            }
+        }
+        return (0.95 * this.sketch.width) / (array_length + 1)
+    }
+
+    // Create a number that represents how many pixels high
+    // each increase of one in the bin array should be
+    height(): number {
+        const binFactorArray = this.binFactorArray()
+        let greatest_value = binFactorArray[0]
+        for (let i = 0; i < binFactorArray.length; i++) {
+            if (binFactorArray[i] > greatest_value) {
+                greatest_value = binFactorArray[i]
+            }
+        }
+        return (0.95 * this.sketch.width) / greatest_value
+    }
+
+    draw() {
+        //these numbers provide the rgb values for the background color
+        this.sketch.background(176, 227, 255)
+        const height = this.height()
+        const binWidth = this.binWidth()
+        const binFactorArray = this.binFactorArray()
+        this.sketch.line(
+            0.025 * this.sketch.width,
+            10,
+            0.025 * this.sketch.width,
+            this.sketch.height
+        )
+        this.sketch.line(
+            0,
+            0.975 * this.sketch.height,
+            this.sketch.width,
+            0.975 * this.sketch.height
+        )
+
+        for (let i = 0; i < binFactorArray.length; i++) {
+            this.sketch.rect(
+                0.025 * this.sketch.width + binWidth * i,
+                0.975 * this.sketch.height - height * binFactorArray[i],
+                binWidth,
+                height * binFactorArray[i]
+            )
+            if (this.binSize != 1) {
+                this.sketch.text(
+                    this.binSize * i
+                        + ' - '
+                        + (i + this.binSize * i - i + this.binSize - 1),
+                    20 + (binWidth * (i + 1) - binWidth / 2),
```

Comment:
If I am reading it correctly, this expression still seems to have an `i` and a `-i` in it, which should cancel, and then it looks like you can factor a this.binSize out of part of it, too...